traffic_ops/traffic_ops_golang/server/servers_assignment.go

```diff
@@ -62,77 +62,141 @@ func getConfigFile(prefix string, xmlId string) string {
 		return
 	} else if !ok {
-		api.HandleErr(w, r, inf.Tx.Tx, http.StatusNotFound, errors.New("no server with that ID found"), nil)
+		api.HandleErr(w, r, tx, http.StatusNotFound, errors.New("no server with that ID found"), nil)
 		return
 	}
 
 	if !strings.HasPrefix(serverInfo.Type, tc.OriginTypeName) {
-		usrErr, sysErr, status := ValidateDSCapabilities(dsList, serverInfo.HostName, inf.Tx.Tx)
+		usrErr, sysErr, status := ValidateDSCapabilities(dsList, serverInfo.HostName, tx)
 		if usrErr != nil || sysErr != nil {
-			api.HandleErr(w, r, inf.Tx.Tx, status, usrErr, sysErr)
+			api.HandleErr(w, r, tx, status, usrErr, sysErr)
 			return
 		}
 	}
 
 	// We already know the CDN exists because that's part of the serverInfo query above
-	serverCDN, _, err := dbhelpers.GetCDNNameFromID(inf.Tx.Tx, int64(serverInfo.CDNID))
+	serverCDN, _, err := dbhelpers.GetCDNNameFromID(tx, int64(serverInfo.CDNID))
 	if err != nil {
 		sysErr = fmt.Errorf("Failed to get CDN name from ID: %v", err)
 		errCode = http.StatusInternalServerError
-		api.HandleErr(w, r, inf.Tx.Tx, errCode, nil, sysErr)
+		api.HandleErr(w, r, tx, errCode, nil, sysErr)
 		return
 	}
 
 	if len(dsList) > 0 {
-		if errCode, userErr, sysErr = checkTenancyAndCDN(inf.Tx.Tx, string(serverCDN), server, serverInfo, dsList, inf.User); userErr != nil || sysErr != nil {
-			api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		if errCode, userErr, sysErr = checkTenancyAndCDN(tx, string(serverCDN), server, serverInfo, dsList, inf.User); userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, tx, errCode, userErr, sysErr)
 			return
 		}
 		if strings.HasPrefix(serverInfo.Type, tc.OriginTypeName) {
-			if userErr, sysErr, status := checkOriginInTopologies(inf.Tx.Tx, serverInfo.Cachegroup, dsList); userErr != nil || sysErr != nil {
-				api.HandleErr(w, r, inf.Tx.Tx, status, userErr, sysErr)
+			if userErr, sysErr, status := checkOriginInTopologies(tx, serverInfo.Cachegroup, dsList); userErr != nil || sysErr != nil {
+				api.HandleErr(w, r, tx, status, userErr, sysErr)
 				return
 			}
 		}
 	}
 
-	assignedDSes, err := assignDeliveryServicesToServer(server, dsList, replace, inf.Tx.Tx)
+	if replace && (serverInfo.Status == tc.CacheStatusOnline.String() || serverInfo.Status == tc.CacheStatusReported.String()) {
+		currentDSIDs, err := checkForLastServerInActiveDeliveryServices(server, dsList, tx)
+		if err != nil {
+			sysErr = fmt.Errorf("checking for deliveryservices to which server #%d is the last assigned: %v", server, err)
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, sysErr)
+			return
+		}
+		if len(currentDSIDs) > 0 {
+			alerts := make([]tc.Alert, 0, len(currentDSIDs))
+			for _, dsID := range currentDSIDs {
+				alert := tc.Alert{
+					Level: tc.ErrorLevel.String(),
+					Text:  fmt.Sprintf("Delivery Service assignment would leave Delivery Service #%d without any assigned ONLINE or REPORTED servers", dsID),
```

Comment:
error message should specify _active_ delivery service, also: `tc.CacheStatusOnline` and `tc.CacheStatusReported`